### PR TITLE
Add restart command

### PR DESCRIPTION
**Pull request by sambazley**
Add ability to restart the window manager, without logging out to the display manager. Can be used, for example, to restart after an update.

### CODE REVIEW

**Comment by baskerville** [comment]
I guess I will remove `examples/loop` after merging this.

File: src/bspwm.c [hunk 1]
```diff
@@ -193,6 +204,25 @@ int main(int argc, char *argv[])
...
+		char *host = NULL;
+		int dn = 0, sn = 0;
+		if (xcb_parse_display(NULL, &host, &dn, &sn) != 0) {
+			snprintf(restart_state_path, sizeof(restart_state_path), RESTART_STATE_PATH_TPL, host, dn, sn);
+		}
+		free(host);
+	}
+
+	if (restart) {
+		FILE *f = fopen(restart_state_path, "w");
```
Comment (MyNameIsCosmo) [inline src/bspwm.c]: Maybe if `RESTART_STATE_ENV_VAR` is not set, the temporary file should be cleaned up?

**Comment by MyNameIsCosmo** [comment]
It seems that upon `bspc wm -r`, this will remove the socket at `/tmp/bspwm_0_0-socket` (or similar), causing issues with other programs subscribed to the socket.

Maybe the [sock_fd](https://github.com/baskerville/bspwm/blob/93c6de166d161f04e2cb57a957363f6245c9dbd1/src/bspwm.c#L113) could be passed from the "original" process to the "new" bspwm process, and [sock cleanup](https://github.com/baskerville/bspwm/blob/93c6de166d161f04e2cb57a957363f6245c9dbd1/src/bspwm.c#L197) can be skipped on restart?
I'm unsure how to handle this, but maybe [this will be a good reference](https://stackoverflow.com/questions/2358684/can-i-share-a-file-descriptor-to-another-process-on-linux-or-are-they-local-to-t).

**Comment by baskerville** [comment]
@MyNameIsCosmo Indeed, it would be great.

**Comment by baskerville** [comment]
I think it might be best to pass `sock_fd` as an environment variable to the new process.

Then, `query_tree` needs to be extended to output the built-in rules. Once this is done, we can avoid calling `run_config` when `-s` is set.

We also need to restore the list of subscribers and the pending rules.

**Comment by sambazley** [comment]
Would it be better if we removed the -s option? Instead, we could load the state file if it exists using RESTART_STATE_ENV_VAR, then remove it when we're done with it.

**Comment by baskerville** [comment]
@sambazley It's fine with me.

**Comment by sambazley** [comment]
@baskerville Sorry, to clarify are you saying it's fine the way it currently is, or the change is fine with you?

**Comment by baskerville** [comment]
You can remove the `-s` flag.

**Comment by MyNameIsCosmo** [comment]
I agree on the sock_fd environment variable. 

I added a rule query to query_tree in #916 

How should restoring the list of subscribers and the pending rules occur? Should there be a query to print the subscribers and pending rules, similar to the state? Should they be included in the state query? Does it matter?

**Comment by baskerville** [comment]
@MyNameIsCosmo I'm working on finishing this in a local branch.

**Comment by MyNameIsCosmo** [comment]
Neat! I'll test this out later today.

**Pull request closed without merging**